File: src/lib/family.ts
import { prisma } from '@/lib/prisma';
import { createSignedUrlResolver } from '@/lib/uploads';

export interface FamilyMemberSummary {
  userId: string;
  membershipId: string;
  name: string;
  email: string;
  username: string;
  emailOrUsername: string;
  avatarUrl: string | null;
  role: string;
  joinedAt: string;
  postCount: number;
}

export async function getFamilyMembers(
  familySpaceId: string
): Promise<FamilyMemberSummary[]> {
  const memberships = await prisma.familyMembership.findMany({
    where: { familySpaceId },
    orderBy: {
      createdAt: 'asc',
    },
    include: {
      user: {
        select: {
          id: true,
          name: true,
          email: true,
          username: true,
          avatarStorageKey: true,
          posts: {
            select: {
              id: true,
            },
          },
        },
      },
    },
  });

  const resolveUrl = createSignedUrlResolver();

  return Promise.all(
    memberships.map(async (membership: any) => {
      const fallbackEmail =
        (membership.user as any).emailOrUsername ?? membership.user.username;
      const email = membership.user.email ?? fallbackEmail ?? '';
      const username =
        membership.user.username ??
        (typeof fallbackEmail === 'string'
          ? fallbackEmail.replace(/[^a-zA-Z0-9_]/g, '').slice(0, 30) || 'user'
          : 'user');

      return {
        userId: membership.userId,
        membershipId: membership.id,
        name: membership.user.name,
        email,
        username,
        emailOrUsername: email,
        avatarUrl: await resolveUrl(membership.user.avatarStorageKey),
        role: membership.role,
        joinedAt: membership.createdAt.toISOString(),
        postCount: membership.user.posts.length,
      };
    })
  );
}

export async function removeFamilyMember(
  familySpaceId: string,
  targetUserId: string
): Promise<{ removed: boolean }> {
  const membership = await prisma.familyMembership.findFirst({
    where: {
      familySpaceId,
      userId: targetUserId,
    },
    select: {
      id: true,
      role: true,
    },
  });

  if (!membership) {
    return { removed: false };
  }

  if (membership.role === 'owner') {
    throw new Error('CANNOT_REMOVE_OWNER');
  }

  await prisma.familyMembership.delete({
    where: { id: membership.id },
  });

  return { removed: true };
}

File: apps/api/src/routers/family.py
from fastapi import APIRouter, Depends, Path
from prisma.errors import PrismaError

from ..db import prisma
from ..dependencies import get_current_user
from ..errors import forbidden, internal_error, not_found
from ..permissions import can_remove_member
from ..schemas.auth import UserResponse
from ..utils import is_cuid

router = APIRouter(prefix="/family/members", tags=["family"])


@router.get("")
async def list_members(user: UserResponse = Depends(get_current_user)):
    try:
        memberships = await prisma.familymembership.find_many(
            where={"familySpaceId": user.familySpaceId},
            include={"user": {"include": {"posts": True}}},
            order={"createdAt": "asc"},
        )
        members = [
            {
                "userId": m.userId,
                "membershipId": m.id,
                "name": m.user.name,
                "emailOrUsername": m.user.emailOrUsername,
                "avatarUrl": m.user.avatarUrl,
                "role": m.role,
                "joinedAt": m.createdAt.isoformat(),
                "postCount": len(m.user.posts) if m.user.posts else 0,
            }
            for m in memberships
            if m.user
        ]
        return {"members": members}
    except PrismaError:
        return internal_error("Failed to load members")


@router.delete("/{user_id}")
async def remove_member(
    user_id: str = Path(..., min_length=1), current_user: UserResponse = Depends(get_current_user)
):
    try:
        if not is_cuid(user_id):
            return not_found("Member not found")
        membership = await prisma.familymembership.find_first(
            where={"familySpaceId": current_user.familySpaceId, "userId": user_id},
            include={"user": True},
        )
        if not membership:
            return not_found("Member not found")

        if not can_remove_member(current_user, membership.userId, membership.role):
            return forbidden("You do not have permission to remove this member")

        await prisma.familymembership.delete(where={"id": membership.id})
        return {"message": "Member removed"}
    except PrismaError:
        return internal_error("Failed to remove member")
